java/arcs/core/util/Parser.kt

```diff
@@ -0,0 +1,397 @@
+ * If you want to handle results without worrying about [Failure] cases, you can [map] them
+ * which only invokes the function if the result is a [Success].
+ *
+ * ```
+ * helloResult.map { value, rest -> doSomething(value) }
+ * ```
+ *
+ * but sometimes you want to supply a default.
+ *
+ * ```
+ * helloResult.orElse { Success("default", "") }
+ * ```
+ *
+ * ## Combining Parsers
+ * You can combine two parsers together in sequence and transform them as well. Here, '+' is
+ * used for sequence combination.
+ *
+ * ```
+ * val helloWorld = token("hello") + token(" world").map { (hello, world) ->
+ *   listOf(hello, world)
+ * }
+ *
+ * val helloResult = helloWorld("hello world")
+ * // returns Success(List<String>("hello", " world"), "")
+ * ```
+ *
+ * ### The [AnyOfParser] parallel 'or' combinator
+ * If you want combine two parsers, so that one or the other may success, use the '/' operator.
+ *
+ * ```kotlin
+ * val helloOrFooWorld = (token("hello") / token("foo")) + token("world")
+ * helloOrFooWorld("fooworld") // success!
+ * helloOrFooWorld("helloworld") // success!
+ * helloOrFooWorld("barworld") // failure!
+ * ```
+ *
+ * ### The Many and Optional combinators.
+ * If a parser can succeed 0 or 1 times, use [optional], and if it can succeed 0 or many times,
+ * use [many]. [optional] returns Success(null, rest) if it didn't match, and [many] returns
+ * Success(List<T>, rest) with the matched results.
+ *
+ * ```kotlin
+ * val helloWorld = token("hello") + optional(token(" ")) + token("world")
+ * helloWorld("hello world") // success!
+ * helloWorld("helloworld") // success!
+ * helloWorld("hello  world") // failure! only one space allowed
+ *
+ * val helloWorld2 = token("hello") + many(token(" ")) + token("world")
+ * helloWorld("hello world") // success!
+ * helloWorld("helloworld") // success!
+ * helloWorld("hello   world") // success!!
+ * ```
+ *
+ * ### Ignoring Parser results
+ * Each time you sequence two parsers via '+', you add another argument to the resulting return
+ * value tuple.
+ *
+ * If you have `Parser<T>` + `Parser<S>`, the result is `ParserResult<Pair<T,S>>` and if you have
+ * `Parser<T> + Parser<S> + Parser<R>` the result is `ParserResult<Triple<T,S,R>>`. Often times,
+ * some of these values are from non-relevant surrounding tokens. You can remove these from the
+ * output signature via the [IgnoringParser], which is normally used via the [unaryMinus] helper.
+ *
+ * ```kotlin
+ * val term = token("hello") / token("world")
+ * val array = token("[") + many(term + optional(token(","))) + token("]")
+ * // without IgnoreParser the return result is Triple<String, List<Pair<String, String?>>, String>
+ *
+ * val arrayClean = -token("[") + many(term + -optional(token(","))) + -token("]")
+ * // arrayClean("[hello,world]") returns Success(List<String>("hello", "world"))
+ * ```
+ */
+abstract class Parser<out T>() {
+    operator fun invoke(string: String, pos: Int) = invoke(string, SourcePosition(pos, 0, 0))
```

Comment:
It looks like `pos` would benefit from a default value `0`, don't you think? In vast majority of usages you would want to parse a string you hold I believe. Would simplify tests as well.